### ISSUE TITLE
Column headers become fixed and out of sync on horizontal scroll

### ISSUE THREAD

**Issue by deepfriedbrain** [angular-ui/ng-grid#3649]
Column headers become fixed and go out of sync on horizontal scroll. I'm using the latest unstable release and Chrome 42x. Refer to the screenshot below.

![image](https://cloud.githubusercontent.com/assets/1531669/7901830/1ee8f1c6-07cf-11e5-8e49-9e9078a095fe.png)

**Comment by c0bra** [comment]
Can you give more details on your grid setup so this can be reproduced, i.e. what features you have turned on/off (like filtering which I see), how many columns total, grid css, etc?

Even better, a plunker with the issue reproduced would be wonderful :)

**Comment by deepfriedbrain** [comment]
I can't reproduce this consistently, but it happens quite frequently (I just couldn't figure out a pattern so far). I'm using ui-grid-selection and ui-grid-pinning on my grid. In the grid css, width and height are set to 100%, though it doesn't really expand the grid to 100% available screen height (that might be another issue). I have 12 columns in total. Needless to say, I have more columns than what can fit into the screen width and hence the need for horizontal scrolling. My grid has these options set on it.
```
		selectedItems: [],
		showGridFooter: false,
		enableRowHeaderSelection: true,
		multiSelect: false,
		enableHorizontalScrollbar: 1,
		enableVerticalScrollbar: 1,
		enableFiltering: true,
		enableSorting: true,
```

If you cannot reproduce it at all, let me know. I'll try to setup a plunker and see if I can reproduce it there.

Thanks to you and all the contributors for the awesome work!

**Comment by deepfriedbrain** [comment]
After playing around a bit, I figured out that 2 things when combined cause the problem:

1. Enable "ui-grid-selection". 
2. Have a div with ng-show around the div for the grid, like so:

```
<div ng-show="tdrs.length > 0">
	<div ui-grid="gridOptions" ui-grid-selection class="tdr-grid"></div>
        <!-- some more content here -->
</div>
```

If I replace ng-show with ng-if, the issue goes away. The issue also goes away if I remove ui-grid-selection from the grid.

I removed every option from the grid, and brought it to down to just 2 columns with a single row of hard-coded data. See the screenshot below. When I first load the page, grid works just fine. But if I refresh the page using browser's refresh button, that's when the header gets frozen. If I navigate to another page (ui router state) and come back to the original page with the grid, it starts to work fine again. But if I refresh the browser again, the issue comes back. So, it happens only after refreshing the browser.

I tried to reproduce it in plunkr but was unsuccessful.

![image](https://cloud.githubusercontent.com/assets/1531669/7986961/4ca34fda-0b0d-11e5-8468-50f73654cc7d.png)

For now my workaround is to use ng-if instead of ng-show.

**Comment by darionato** [comment]
I have the same issue. With hg-if it fix it but for me is too slow.
If I resize the chrome tab it starts to scroll.
I'm using the following directives on my grid: ui-grid-resize-columns, ui-grid-pagination, ui-grid-pinning, ui-grid-selection, ui-grid-edit, ui-grid-row-edit

**Comment by c0bra** [comment]
@deepfriedbrain Did you put that exact app in plunker? And did you try it in the full browser view rather than in the iframe with the editor?

**Comment by deepfriedbrain** [comment]
Yes I tried to use the full browser but still couldn't reproduce it in a plunker.

**Comment by darionato** [comment]
I created a Plunker to show this strange behavior: Here the url:

http://plnkr.co/edit/7vF92f9VR8Ck5OYDfHBd?p=preview

When you run the project, ensure that the preview is smaller than the grid so the horizontal scroll is active. When you scroll the header stay locked.

**Comment by JLLeitschuh** [comment]
In chrome I can't even scroll in your example. I have to use the arrow keys to scroll to the right/left. At that point it is very obvious that the headers aren't moving.

**Comment by darionato** [comment]
I tried this example in chrome and firefox and I can scroll left-right.

When the example starts, shrink the preview column quite small, press the button Show/hide and the grid should appear with the scroll bar in the bottom.

**Comment by dprothero** [comment]
I've encountered the same behavior and the Plunkr that @darionato created does exhibit the behavior. Switching to using an ng-if instead of ng-show for now.

**Comment by justin-lovell** [comment]
I ran into this issue, but this was primarily due to the grid being hosted in a Bootstrap tab.

The duplication is that the grid is initially hidden since it is located on a non-default tab. The root cause is that ui-grid does not know how to calculate the column alignments, due to the grid being not visible in the DOM.

Funny enough, [the fix has already been documented](http://ui-grid.info/docs/#/tutorial/110_grid_in_modal). The key is:

 ```
$rootScope.gridOptions = {
    onRegisterApi: function (gridApi) {
      $scope.gridApi = gridApi;
 
      // call resize every 200 ms for 2 s after modal finishes opening - usually only necessary on a bootstrap modal
      $interval( function() {
        $scope.gridApi.core.handleWindowResize();
      }, 10, 500);
      }
  };
```

**Comment by justin-lovell** [comment]
Further to the above, the auto-resize module implies the same behaviour as the solution: http://plnkr.co/edit/fwdXMamTBrR1yKVwcohZ?p=preview

**Comment by darionato** [comment]
thank you, with this trick it works fine

**Comment by angular-git-user** [comment]
Thanks the above worked for me!!!

**Comment by mgrivera** [comment]
Please read the angular docs for $interval. If you are going to execute it indefinitely, ie: with no definite number of times, you must destroy $interval, maybe when $scope is destroyed.

**Comment by saadgit** [comment]
In addition to whats said above as solution, what worked for me was, updated all the dependencies and moved to the fresh release and set the css property of 'overflow' to 'hidden' (instead of 'visible') when using this class "ui-grid-top-panel" for grid header.

**Comment by pawanupr16** [comment]
I have facing same issue in this when we click on tab button  on search box on each column in grid ui on top of  the header then missing alignment of the header and  table data 
bellow image is there 
please help me to solve this issue 

![test](https://cloud.githubusercontent.com/assets/22906190/20589241/48635a5a-b241-11e6-8276-0a80e908d20e.png)

**Comment by pawanupr16** [comment]
Any updates on this ??

Thanks in advance

**Comment by prashant93** [comment]
iIsue of ng-show still exists in ui-grid but @justin-lovell  trick help me out. Thanks for this bro.

**Comment by jasonbrandt42** [comment]
I am also having this issue.  100% reproducible on a grid where we set the horizontalscroll to 1 after the grid was initialized with horizontalscroll of 0.  

headers become fixed and do not scroll with the data.

**Comment by giltwg** [comment]
This works for me:
ui-grid-header-viewport = hidden !important;

**Comment by AThomsonTrimble** [comment]
I still get this 100% of the time when initial grid size is wider than browser window.  Doing any kind of resize (either zooming or click+drag window sides) causes the issue to go away.  None of the above workarounds were effective.

**Comment by sohail30** [comment]
Hi, even I am facing the same problem for my grid headers are fixed. doing any kind of resize causes the issue goes away. but we cant rely on it can anyone help me  to solve this issue??

**Comment by WorkMAE** [comment]
This is happening for me when I load a grid inside a modal, I have tried all of the fixes here and cannot get it working properly.  The headers become fixed and no longer scroll with the rest of the grid horizontally.
